src/main/kotlin/cacophony/parser/LLOneParser.kt

```diff
@@ -17,6 +28,60 @@ class LLOneParser<SymbolType : Enum<SymbolType>> : Parser<SymbolType> {
         terminals: List<ParseTree.Leaf<SymbolType>>,
         diagnostics: Diagnostics,
     ): ParseTree<SymbolType> {
-        TODO("Not yet implemented")
+        val terminalIterator = terminals.iterator()
+        var terminal = terminalIterator.next()
+        var eof = false
+
+        fun goToSyncSymbol() {
+            while (!syncSymbols.contains(terminal.token.category) && terminalIterator.hasNext()) {
+                terminal = terminalIterator.next()
+            }
+        }
+
+        fun topDownParse(symbol: SymbolType): ParseTree<SymbolType> {
+            if (symbol == terminal.token.category) {
+                return terminal.also {
+                    if (terminalIterator.hasNext()) {
+                        terminal = terminalIterator.next()
+                    } else {
+                        eof = true
+                    }
+                }
+            }
+
+            var children = mutableListOf<ParseTree<SymbolType>>()
+
+            val dfa = automata[symbol]!!
+            var state = dfa.getStartingState()
+
+            do {
+                nextAction[terminal.token.category]?.get(Pair(state, dfa))?.let { nextSymbol ->
+                    dfa.getProduction(state, nextSymbol)?.let { nextState ->
+                        try {
+                            state = nextState
+                            children.add(topDownParse(nextSymbol))
+                        } catch (e: ParserError) {
+                            if (dfa.getProduction(state, terminal.token.category) == null) {
```

Comment:
on this line I have red error `'if' must have both main and 'else' branches if used as an expression`

Reply:
I do as well but to me it looks like the LSP hallucinating